README.md
**_Important Notice:_**
Due to a [change in the AWS Lambda execution environment](https://aws.amazon.com/blogs/compute/upcoming-updates-to-the-aws-lambda-execution-environment/), Serverless Image Handler v3 deployments are functionally broken. To address the issue we have released [minor version update v3.1.1](https://solutions-reference.s3.amazonaws.com/serverless-image-handler/v3.1.1/serverless-image-handler.template). We recommend all users of v3 to run cloudformation stack update with v3.1.1. Additionally, we suggest you to look at v5 of the solution and migrate to v5 if it addresses all of your use cases.

# AWS Serverless Image Handler Lambda wrapper for SharpJS
A solution to dynamically handle images on the fly, utilizing Sharp (https://sharp.pixelplumbing.com/en/stable/).
Published version, additional details and documentation are available here: https://aws.amazon.com/solutions/serverless-image-handler/

_Note:_ it is recommended to build the application binary on Amazon Linux.

## On This Page
- [Architecture Overview](#architecture-overview)
- [Creating a custom build](#creating-a-custom-build)
- [External Contributors](#external-contributors)
- [License](#license)

## Architecture Overview
![Architecture](architecture.png)

The AWS CloudFormation template deploys an Amazon CloudFront distribution, Amazon API Gateway REST API, and an AWS Lambda function. Amazon CloudFront provides a caching layer to reduce the cost of image processing and the latency of subsequent image delivery. The Amazon API Gateway provides endpoint resources and triggers the AWS Lambda function. The AWS Lambda function retrieves the image from the customer's Amazon Simple Storage Service (Amazon S3) bucket and uses Sharp to return a modified version of the image to the API Gateway. Additionally, the solution generates a CloudFront domain name that provides cached access to the image handler API.

_**Note**:_ From v5.0, all AWS CloudFormation template resources are created be [AWS CDK](https://aws.amazon.com/cdk/) and [AWS Solutions Constructs](https://aws.amazon.com/solutions/constructs/). Since the AWS CloudFormation template resources have the same logical ID comparing to v4.x, it makes the solution upgradable mostly from v4.x to v5.

## Creating a custom build
The solution can be deployed through the CloudFormation template available on the solution home page.
To make changes to the solution, download or clone this repo, update the source code and then run the deployment/build-s3-dist.sh script to deploy the updated Lambda code to an Amazon S3 bucket in your account.

### Prerequisites:
* [AWS Command Line Interface](https://aws.amazon.com/cli/)
* Node.js 12.x or later

### 1. Clone the repository
```bash
git clone https://github.com/awslabs/serverless-image-handler.git
```

### 2. Run unit tests for customization
Run unit tests to make sure added customization passes the tests:
```bash
cd ./deployment
chmod +x ./run-unit-tests.sh
./run-unit-tests.sh
```

### 3. Declare environment variables
```bash
export REGION=aws-region-code # the AWS region to launch the solution (e.g. us-east-1)
export DIST_OUTPUT_BUCKET=my-bucket-name # bucket where customized code will reside
export SOLUTION_NAME=my-solution-name # the solution name
export VERSION=my-version # version number for the customized code
```

### 4. Create an Amazon S3 Bucket
The CloudFormation template is configured to pull the Lambda deployment packages from Amazon S3 bucket in the region the template is being launched in. Create a bucket in the desired region with the region name appended to the name of the bucket.
```bash
aws s3 mb s3://$DIST_OUTPUT_BUCKET-$REGION --region $REGION
```

### 5. Create the deployment packages
Build the distributable:
```bash
chmod +x ./build-s3-dist.sh
./build-s3-dist.sh $DIST_OUTPUT_BUCKET $SOLUTION_NAME $VERSION
```

Deploy the distributable to the Amazon S3 bucket in your account:
```bash
aws s3 sync ./regional-s3-assets/ s3://$DIST_OUTPUT_BUCKET-$REGION/$SOLUTION_NAME/$VERSION/ --acl bucket-owner-full-control
aws s3 sync ./global-s3-assets/ s3://$DIST_OUTPUT_BUCKET-$REGION/$SOLUTION_NAME/$VERSION/ --acl bucket-owner-full-control
```

### 6. Launch the CloudFormation template.
* Get the link of the `serverless-image-handler.template` uploaded to your Amazon S3 bucket.
* Deploy the Serverless Image Handler solution to your account by launching a new AWS CloudFormation stack using the S3 link of the `serverless-image-handler.template`.

## External Contributors
- [@leviwilson](https://github.com/leviwilson) for [#117](https://github.com/awslabs/serverless-image-handler/pull/117)
- [@rpong](https://github.com/rpong) for [#130](https://github.com/awslabs/serverless-image-handler/pull/130)
- [@harriswong](https://github.com/harriswong) for [#138](https://github.com/awslabs/serverless-image-handler/pull/138)
- [@ganey](https://github.com/ganey) for [#139](https://github.com/awslabs/serverless-image-handler/pull/139)
- [@browniebroke](https://github.com/browniebroke) for [#151](https://github.com/awslabs/serverless-image-handler/pull/151), [#152](https://github.com/awslabs/serverless-image-handler/pull/152)
- [@john-shaffer](https://github.com/john-shaffer) for [#158](https://github.com/awslabs/serverless-image-handler/pull/158)
- [@toredash](https://github.com/toredash) for [#174](https://github.com/awslabs/serverless-image-handler/pull/174), [#195](https://github.com/awslabs/serverless-image-handler/pull/195)
- [@lith-imad](https://github.com/lith-imad) for [#194](https://github.com/awslabs/serverless-image-handler/pull/194)
- [@pch](https://github.com/pch) for [#227](https://github.com/awslabs/serverless-image-handler/pull/227)
- [@atrope](https://github.com/atrope) for [#201](https://github.com/awslabs/serverless-image-handler/pull/201)
- [@bretto36](https://github.com/bretto36) for [#182](https://github.com/awslabs/serverless-image-handler/pull/182)
- [@makoncline](https://github.com/makoncline) for [#255](https://github.com/awslabs/serverless-image-handler/pull/255)


## Saving output image to S3

It is possible to save the output image to a S3 bucket adding extra HTTP Headers.
In order to enable output saving:

- Open the AWS console and go to the Lambda section
- Open the *-ImageHandlerFunction-* lambda
- Add S3_SAVE_SECRET to the Lambda environment. Choose a secret key long enough, e.g. 32 characters, digits and letters in mixed
  case. This is the key that must be supplied by the invoker to save on S3.
- (optional) add S3_SAVE_BUCKET to the Lambda environment. If missing, TC_AWS_LOADER_BUCKET is used as fallback.
  This is the target bucket.
- edit the Lambda role to add "s3:PutObject" permission to the policy on the target bucket (S3_SAVE_BUCKET).
- save the changes

Now you are ready to request save output. You need to set `X-save-s3-key` to the request header with the value of the
S3 path (key) where the file must be saved. In this case you must also set the `X-save-secret` request header with a string
that matches the S3_SAVE_SECRET.

Now you can issue two different type of requests, depending on your needs:

- GET request: the output image to be returned in the body of the request. HTTP status will be 200 OK;
- POST request: the output image to be returned in the body of the request. HTTP status will be 201 Created.

Example:

    # image returned in the body
	curl -H "X-save-s3-key: my/output/image.jpg" -H "X-save-secret: 123456789" https://a12345.execute-api.us-east-1.amazonaws.com/image/300x300/source.jpg

    # image not returned in the body
	curl -I -X POST -H "X-save-s3-key: out/image.jpg" -H "X-save-secret: ayzd5BEk7znTCMgNRQMkfyfTa54A6vA6" https://a12345.execute-api.us-east-1.amazonaws.com/image/300x300/source.jpg

## Debugging on Windows & IntelliJ idea with AWS Toolkit
After using `build-s3-dist.sh` make sure to delete node_modules from source/image-handler and run `npm install` to reinstall dependencies for the correct platform.

## Request example
Requests for the image-handler must be encoded in base64.
Thumbor-based urls can still be used, however it is only possible to resize the image. Adding cropping to the url will make the request fail.

A standard, yet to be encoded, request is:

```
{
    "bucket": "bucket-name",
    "key": "image-name",
    "edits":{
        "extract":{
            "left":100,
            "top":100,
            "width":300,
            "height":200
        },
        "resize":{
            "width":100
        },
        "smartCrop": true 
     }
}
```

Any mistake in the definition of the request object will most likely result in one of the following errors:
 
```
ERROR    {
  status: 404,
  code: 'NoSuchKey',
  message: 'The specified key does not exist.'
}
```

or 

```
ERROR	{
  status: 400,
  code: 'DecodeRequest::CannotDecodeRequest',
  message: 'The image request you provided could not be decoded. Please check that your request is base64 encoded properly and refer to the documentation for additional guidance.'
}
```

For more options, refer to the [sharp documentation](https://sharp.pixelplumbing.com/). 
***
## License
Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.<br />
SPDX-License-Identifier: Apache-2.0


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [5.2.0] - 2021-01-29
### Added
- Support for ap-east-1 and me-south-1 regions: [#192](https://github.com/awslabs/serverless-image-handler/issues/192), [#228](https://github.com/awslabs/serverless-image-handler/issues/228), [#232](https://github.com/awslabs/serverless-image-handler/issues/232)
- Unit tests for custom-resource: `100%` coverage
- Cloudfront cache policy and origin request policy: [#229](https://github.com/awslabs/serverless-image-handler/issues/229)
- Circular cropping feature: [#214](https://github.com/awslabs/serverless-image-handler/issues/214), [#216](https://github.com/awslabs/serverless-image-handler/issues/216)
- Unit tests for image-handler: `100%` coverage
- Support for files without extension on thumbor requests: [#169](https://github.com/awslabs/serverless-image-handler/issues/169), [#188](https://github.com/awslabs/serverless-image-handler/issues/188)
- Inappropriate content detection feature: [#243](https://github.com/awslabs/serverless-image-handler/issues/243)
- Unit tests for image-request: `100%` coverage

### Fixed
- Graceful failure when no faces are detected using smartCrop and fail on resizing before smartCrop: [#132](https://github.com/awslabs/serverless-image-handler/issues/132), [#133](https://github.com/awslabs/serverless-image-handler/issues/133)
- Broken SVG returned if no edits specified and Auto-WebP enabled: [#247](https://github.com/awslabs/serverless-image-handler/issues/247)
- Removed "--recursive" from README.md: [#255](https://github.com/awslabs/serverless-image-handler/pull/255)
- fixed issue with failure on resize if width or height is float: [#254](https://github.com/awslabs/serverless-image-handler/issues/254)

### Changed 
- Constructs test template for constructs unit test: `100%` coverage

## [5.1.0] - 2020-11-19
### ⚠ BREAKING CHANGES
- **Image URL Signature**: When image URL signature is enabled, all URLs including existing URLs should have `signature` query parameter.

### Added
- Image URL signature: [#111](https://github.com/awslabs/serverless-image-handler/issues/111), [#203](https://github.com/awslabs/serverless-image-handler/issues/203), [#221](https://github.com/awslabs/serverless-image-handler/issues/221), [#227](https://github.com/awslabs/serverless-image-handler/pull/227)
- AWS Lambda `413` error handling. When the response payload is bigger than 6MB, it throws `TooLargeImageException`: [#35](https://github.com/awslabs/serverless-image-handler/issues/35), [#97](https://github.com/awslabs/serverless-image-handler/issues/97), [#193](https://github.com/awslabs/serverless-image-handler/issues/193), [#204](https://github.com/awslabs/serverless-image-handler/issues/204)
- Default fallback image: [#137](https://github.com/awslabs/serverless-image-handler/issues/137)
- Unit tests for custom resource: `100%` coverage
- Add `SVG` support. When any edits are used, the output would be automatically `PNG` unless the output format is specified: [#31](https://github.com/awslabs/serverless-image-handler/issues/31), [#234](https://github.com/awslabs/serverless-image-handler/issues/234)
- Custom headers: [#182](https://github.com/awslabs/serverless-image-handler/pull/182)
- Enabling ALB Support : [#201](https://github.com/awslabs/serverless-image-handler/pull/201)

### Fixed
- Thumbor paths broken if they include "-" and "100x100": [#208](https://github.com/awslabs/serverless-image-handler/issues/208)
- Rewrite doesn't seem to be working: [#121](https://github.com/awslabs/serverless-image-handler/issues/121)
- Correct EXIF: [#197](https://github.com/awslabs/serverless-image-handler/issues/197), [#220](https://github.com/awslabs/serverless-image-handler/issues/220), [#235](https://github.com/awslabs/serverless-image-handler/issues/235), [#236](https://github.com/awslabs/serverless-image-handler/issues/236), [#240](https://github.com/awslabs/serverless-image-handler/issues/240)
- Sub folder support in Thumbor `watermark` filter: [#231](https://github.com/awslabs/serverless-image-handler/issues/231)

### Changed
- AWS CDK and AWS Solutions Constructs version (from 1.57.0 to 1.64.1)
- sharp base version (from 0.25.4 to 0.26.1)
- Refactors the custom resource Lambda source code
- Migrate unit tests to use `jest`
- Move all `aws-sdk` in `ImageHandler` Labmda function to `index.js` for the best practice
- Enhance the default error message not to show empty JSON: [#206](https://github.com/awslabs/serverless-image-handler/issues/206)

### Removed
- Remove `manifest-generator`

## [5.0.0] - 2020-08-31
### Added
- AWS CDK and AWS Solutions Constructs to create AWS CloudFormation template

### Fixed
- Auto WebP does not work properly: [#195](https://github.com/awslabs/serverless-image-handler/pull/195), [#200](https://github.com/awslabs/serverless-image-handler/issues/200), [#205](https://github.com/awslabs/serverless-image-handler/issues/205)
- A bug where base64 encoding containing slash: [#194](https://github.com/awslabs/serverless-image-handler/pull/194)
- Thumbor issues:
  - `0` size support: [#183](https://github.com/awslabs/serverless-image-handler/issues/183)
  - `convolution` filter does not work: [#187](https://github.com/awslabs/serverless-image-handler/issues/187)
  - `fill` filter does not work: [#190](https://github.com/awslabs/serverless-image-handler/issues/190)
- __Note that__ duplicated features has been merged gracefully.

### Removed
- AWS CloudFormation template: `serverless-image-handler.template`

### Changed
- sharp base version (from 0.23.4 to 0.25.4)
- Remove `Promise` to return since `async` functions return promises: [#189](https://github.com/awslabs/serverless-image-handler/issues/189)

source/image-handler/lambda_function.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
##############################################################################
#  Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.   #
#                                                                            #
#  Licensed under the Amazon Software License (the "License"). You may not   #
#  use this file except in compliance with the License. A copy of the        #
#  License is located at                                                     #
#                                                                            #
#      http://aws.amazon.com/asl/                                            #
#                                                                            #
#  or in the "license" file accompanying this file. This file is distributed #
#  on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,        #
#  express or implied. See the License for the specific language governing   #
#  permissions and limitations under the License.                            #
##############################################################################

from __future__ import print_function
import cStringIO
import base64
import requests_unixsocket
import threading
import time
import mimetypes
import traceback
import os.path
import json
import os
import timeit
import ast
from image_handler import lambda_metrics
from image_handler import lambda_rewrite
from PIL import Image
from io import BytesIO
from distutils.util import strtobool
import json
from hashlib import sha1
import hmac
import base64
import re
from requests.compat import quote

from tornado.httpserver import HTTPServer
from tornado.netutil import bind_unix_socket
from tornado.options import options, define

from thumbor.context import ServerParameters
from thumbor.server import *

import boto3

thumbor_config_path = '/var/task/image_handler/thumbor.conf'
thumbor_socket = '/tmp/thumbor'
unix_path = 'http+unix://%2Ftmp%2Fthumbor'

##############################################################################
# helper methods
#
def response_formater(status_code='400',
                      body={'message': 'error, please check lambda logs'},
                      cache_control='max-age=120,public',
                      content_type='application/json',
                      expires='',
                      etag='',
                      date='',
                      vary=False
                      ):

    api_response = {
        'statusCode': status_code,
        'headers': {
            'Content-Type': content_type
        }
    }

    if str(os.environ.get('ENABLE_CORS')).upper() == "YES":
        api_response['headers']['Access-Control-Allow-Origin'] = os.environ.get('CORS_ORIGIN')

    # SO-SIH-175 - 08/28/2018 - Missing header
    # Adding missing header to response
    # https://github.com/awslabs/serverless-image-handler/pull/34
    # https://github.com/awslabs/serverless-image-handler/pull/60
    if int(status_code) != 200:
        api_response['body'] = json.dumps(body)
        api_response['headers']['Cache-Control'] = cache_control
        api_response['isBase64Encoded'] = 'false'
    else:
        api_response['body'] = body
        api_response['isBase64Encoded'] = 'true'
        api_response['headers']['Expires'] = expires
        api_response['headers']['Etag'] = etag
        api_response['headers']['Cache-Control'] = cache_control
        api_response['headers']['Date'] = date
    if vary:
        api_response['headers']['Vary'] = vary
    logging.debug('api response: %s' % (api_response))
    return api_response

def auto_webp(original_request, request_headers):
    headers = {'Accept':'*/*'}
    vary = bool(strtobool(str(config.AUTO_WEBP)))
    if vary:
        if original_request.get('headers'):
            if original_request['headers'].get('Accept'):
                request_headers['Accept'] = original_request['headers']['Accept']
    return vary, request_headers

# SO-SIH-166 - 08/08/2018 - Enabling safe url
# Encoding url and hashing with security key
def encoding_string(string):
    """
    Encoding URL per RFC 3986.
    """
    return(quote(string,safe=''))

def signed_url(secret_key, string_to_sign):
    """
    Signing URL with security key
    """
    hashed = hmac.new(secret_key,string_to_sign, sha1)
    return base64.b64encode(hashed.digest())


def true_url(http_path):
    """
    Generate URL based on /unsafe or security key
    """
    if bool(strtobool(str(config.ALLOW_UNSAFE_URL))):
        http_path = '/unsafe' + http_path
    return http_path


def rewrite(http_path):
    if str(os.environ.get('REWRITE_ENABLED')).upper() == 'YES':
        http_path = lambda_rewrite.match_patterns(http_path)
    return http_path

def gen_body(ctype, content):
    """
    Convert image to base64 to be sent as body response.
    """
    try:
        format_ = ctype[ctype.find('/')+1:]
        supported = ['jpeg', 'png', 'gif']
        if format_ not in supported:
            None
        return base64.b64encode(content)
    except Exception as error:
        logging.error('gen_body error: %s' % (error))
        logging.error('gen_body trace: %s' % traceback.format_exc())
        return None

def send_metrics(event, result, start_time):
    """
    Send anonymous usage metrics to AWS.
    """
    t = threading.Thread(
        target=lambda_metrics.send_data,
        args=(event, result, start_time, )
    )
    t.start()
    return t

##############################################################################
# server methods
#
def run_server(application, context):
    server = HTTPServer(application)
    define(
        'unix_socket',
        group='webserver',
        default=thumbor_socket,
        help='Path to unix socket to bind')
    socket = bind_unix_socket(options.unix_socket)
    server.add_socket(socket)
    server.start(1)

def stop_thumbor():
    return None
    tornado.ioloop.IOLoop.instance().stop()
    try:
        os.remove(thumbor_socket)
    except OSError as error:
        logging.error('stop_thumbor error: %s' % (error))

def start_thumbor():
    """
    Runs thumbor server with the specified arguments.
    """
    try:
        server_parameters = ServerParameters(
            port=8888,
            ip='0.0.0.0',
            config_path=None,
            keyfile=False,
            log_level=log_level,
            app_class='thumbor.app.ThumborServiceApp',
            use_environment=True)
        global config
        # SO-SIH-167 - 08/08/2018 - Allowing environment variable
        # Passing environment variable flag to server parameters
        config = get_config(thumbor_config_path, server_parameters.use_environment)
        configure_log(config, server_parameters.log_level)
        importer = get_importer(config)
        os.environ["PATH"] += os.pathsep + '/var/task'
        validate_config(config, server_parameters)
        with get_context(server_parameters, config, importer) as thumbor_context:
            application = get_application(thumbor_context)
            run_server(application, thumbor_context)
            tornado.ioloop.IOLoop.instance().start()
            logging.info(
                'thumbor running at %s:%d' %
                (thumbor_context.server.ip, thumbor_context.server.port)
            )
            return config
    except RuntimeError as error:
        if str(error) != "IOLoop is already running":
            logging.error('start_thumbor RuntimeError: %s' % (error))
            stop_thumbor()
    except Exception as error:
        stop_thumbor()
        logging.error('start_thumbor error: %s' % (error))
        logging.error('start_thumbor trace: %s' % traceback.format_exc())

def start_server():
    t = threading.Thread(target=start_thumbor)
    t.daemon = True
    t.start()
    return t

def restart_server():
    threads = threading.enumerate()
    main_thread = threading.current_thread()
    for t in threads:
        if t is not main_thread:
            t.exit()
            t.join()
    start_server()

##############################################################################
# request processing methods
#
def is_thumbor_down():
    if not os.path.exists(thumbor_socket):
        start_server()
    session = requests_unixsocket.Session()
    http_health = '/healthcheck'
    retries = 10
    while(retries > 0):
        try:
            response = session.get(unix_path + http_health)
            if (response.status_code == 200):
                break
        except Exception as error:
            time.sleep(0.03)
            retries -= 1
            continue
    if retries <= 0:
        logging.error(
            'call_thumbor error: tornado server unavailable,\
            proceeding with tornado server restart'
        )
        restart_server()
        return response_formater(status_code='502')
    return False, session

def request_thumbor(original_request, session):
    """
    Uses requests_unixsocket to send http
    requests over unix domain socket/thumbor.
    """
    logging.debug('original_request: %s' % (json.dumps(original_request)))
    http_path = original_request['path']
    logging.debug('original_request path: %s' % (http_path))
    try:
        http_path = rewrite(http_path);
        logging.debug('http path after rewrite: %s' % (http_path))
        http_path = true_url(http_path)
    except Exception as error:
        logging.error('invalid http path: %s' % (error))
    request_headers = {}
    vary, request_headers = auto_webp(original_request, request_headers)
    return session.get(unix_path + http_path, headers=request_headers), vary

def process_thumbor_responde(thumbor_response, vary, original_request):
    if thumbor_response.status_code != 200:
        return response_formater(status_code=thumbor_response.status_code)
    if vary:
        vary = thumbor_response.headers['vary']
    content_type = thumbor_response.headers['content-type']

    # Save output to S3 if x-save-s3-key header is set.
    if original_request.get('headers'):
        s3_key = original_request['headers'].get('x-save-s3-key')
        logging.debug('Save to S3: %s' % s3_key)

        if s3_key :
            s3_passed_secret = original_request['headers'].get('x-save-secret')

            if s3_passed_secret :
                if s3_passed_secret == os.environ.get('S3_SAVE_SECRET') :
                    if (not os.environ.get('S3_SAVE_BUCKET')) :
                        s3_save_bucket = os.environ.get('TC_AWS_LOADER_BUCKET') # fallback
                    else :
                        s3_save_bucket = os.environ.get('S3_SAVE_BUCKET')

                    logging.debug('Save bucket: %s' % s3_save_bucket)
                    s3 = boto3.resource('s3')
                    object = s3.Object(s3_save_bucket, s3_key)
                    object.put(Body=thumbor_response.content, ContentType=content_type)
                    logging.debug('File saved: %s' % s3_key)

                    # POST method returns HTTP 201 Created and an empty body. GET returns the image in the body.
                    if original_request['requestContext']['httpMethod'] == 'POST' :
                        return response_formater(status_code=201, body={ 'size': int(thumbor_response.headers['content-length']) })
                else :
                    logging.error('Wrong x-save-secret, authentication failed')
                    return response_formater(status_code=404)
            else :
                logging.error('Missing mandatory X-save-secret header')

    body = gen_body(content_type, thumbor_response.content)
    # SO-SIH-173 - 08/20/2018 - Lambda payload limit
    # Lambda limits to 6MB of response payload
    # https://docs.aws.amazon.com/lambda/latest/dg/limits.html
    content_length = int(thumbor_response.headers['content-length'])
    logging.debug('content length: %s' % (content_length))
    if (content_length > 6000000):
        return response_formater(status_code='500',
                                 body={'message': 'body size is too long'},
                                 )
    if body is None:
        return response_formater(status_code='500',
                                 cache_control='no-cache,no-store')
    return response_formater(status_code='200',
                             body=body,
                             cache_control=thumbor_response.headers['Cache-Control'],
                             content_type=content_type,
                             expires=thumbor_response.headers['Expires'],
                             etag=thumbor_response.headers['Etag'],
                             date=thumbor_response.headers['Date'],
                             vary=vary
                             )

def call_thumbor(original_request):
    thumbor_down, session = is_thumbor_down()
    if thumbor_down:
        return thumbor_down
    thumbor_response, vary = request_thumbor(original_request, session)
    return process_thumbor_responde(thumbor_response, vary, original_request)

def lambda_handler(event, context):
    """
    Main event handler, calls thumbor with received event.
    """
    try:
        start_time = timeit.default_timer()
        global log_level
        log_level = str(os.environ.get('LOG_LEVEL')).upper()
        if log_level not in [
            'DEBUG', 'INFO',
            'WARNING', 'ERROR',
            'CRITICAL'
        ]:
            log_level = 'ERROR'
        logging.getLogger().setLevel(log_level)

        if event['requestContext']['httpMethod'] != 'GET' and \
                event['requestContext']['httpMethod'] != 'HEAD' and \
                event['requestContext']['httpMethod'] != 'POST':
            return response_formater(status_code=405)
        result = call_thumbor(event)
        if str(os.environ.get('SEND_ANONYMOUS_DATA')).upper() == 'YES':
            send_metrics(event, result, start_time)
        return result
    except Exception as error:
        logging.error('lambda_handler error: %s' % (error))
        logging.error('lambda_handler trace: %s' % traceback.format_exc())
        return response_formater(status_code='500',
                                 cache_control='no-cache,no-store')
